### PR TITLE
[Ibis] Divorce port name (hints) from port symbol names

### PR DESCRIPTION
**Pull request by mortbopet**
Port symbols now have no effect on the eventual port name of a module. This also implies that multiple ports can have the same namehint, and will be uniqued when lowering `ibis.container` to `hw.module` (see new test in `containers_to_hw.mlir`).

### CODE REVIEW

File: lib/Dialect/Ibis/Transforms/IbisTunneling.cpp [hunk 1]
```diff
@@ -296,7 +296,8 @@ LogicalResult Tunneler::tunnelDown(InstanceGraphNode *currentContainer,
   llvm::DenseMap<StringAttr, OutputPortOp> outputPortOps;
   for (PortInfo &pi : portInfos) {
     outputPortOps[pi.portName] = rewriter.create<OutputPortOp>(
-        op.getLoc(), circt::hw::InnerSymAttr::get(pi.portName), pi.getType());
+        op.getLoc(), pi.portName, circt::hw::InnerSymAttr::get(pi.portName),
```
Comment (teqdruid) [inline lib/Dialect/Ibis/Transforms/IbisTunneling.cpp:299]: Unrelated to this PR, but will `circt::hw::InnerSymAttr::get(pi.portName)` create a unique symbol?

File: lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp [hunk 2]
```diff
@@ -34,41 +35,54 @@ struct ContainerPortInfo {
...
     };
 
-    // Gather in and output port ops.
+    // Gather in and output port ops to define the hw.module interface. Here, we
+    // also perform uniqueing of the port names.
+    Namespace portNs;
     for (auto input : container.getBodyBlock()->getOps<InputPortOp>()) {
-      opInputs[input.getInnerSym().getSymName()] = input;
-
+      auto uniquePortName =
```
Comment (teqdruid) [inline lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp:66]: I may be wrong about this, but I don't think port names are required to be unique in the `hw.module` IR. IIRC ExportVerilog will take care of uniquing port names. In the IR, they are pretty much only name hints -- only order matters.

File: lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #include "circt/Dialect/Ibis/IbisPasses.h"
 #include "circt/Dialect/Ibis/IbisTypes.h"
 
+#include "circt/Support/Namespace.h"
```
Comment (teqdruid) [inline lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp:17]: I don't see where this is used.

File: include/circt/Dialect/Ibis/IbisOps.td [hunk 4]
```diff
@@ -778,6 +785,9 @@ class PortLikeOp<string mnemonic, list<Trait> traits = []> :
   code extraPortClassDefinition = ?;
   let extraClassDefinition =  extraPortClassDefinition # [{
     LogicalResult $cppClass::verify() {
+      if(getName().empty())
+        return emitOpError("port name cannot be empty");
```
Comment (teqdruid) [inline include/circt/Dialect/Ibis/IbisOps.td:789]: Should we default to the symbol name?

**Review by teqdruid — APPROVED** [review]
Do you plan to do the same for containers in a subsequent PR?

File: lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp (hunk 3 above)
Comment (mortbopet) [reply lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp:17]: l. 64 in this file

File: include/circt/Dialect/Ibis/IbisOps.td (hunk 4 above)
Comment (mortbopet) [reply include/circt/Dialect/Ibis/IbisOps.td:789]: In that case, i think the assembly format should be `ibis.port.input sym @foo` - that is, making the namehint optional. We can do that.

Comment (teqdruid) [reply include/circt/Dialect/Ibis/IbisOps.td:789]: I don't feel strongly about it either way.

**Comment by mortbopet** [comment]
> Do you plan to do the same for containers in a subsequent PR?
Yep. I think ports are highest priority for getting fixed, and then containers.

**Pull request merged**